### PR TITLE
Add verify_batch_empty_fail_different_msg

### PR DESCRIPTION
**Pull request by jonas-lj**
Closing #97 
Closing #98 

Note that this PR contains the same changes as #104 and #107.

### CODE REVIEW

File: fastcrypto/src/bls12381.rs [hunk 1]
```diff
@@ -206,23 +206,35 @@ impl VerifyingKey for BLS12381PublicKey {
...
-                "Mismatch between number of signatures and public keys provided"
+                "Mismatch between number of messages, signatures and public keys provided"
             ));
         }
-        let mut rands: Vec<blst_scalar> = Vec::with_capacity(num_sigs);
+        let mut rands: Vec<blst_scalar> = Vec::with_capacity(sigs.len());
         let mut rng = OsRng;
 
-        for _ in 0..num_sigs {
+        for _ in 0..sigs.len() {
```
Comment (kchalkias) [inline fastcrypto/src/bls12381.rs:237]: Unrelated to this PR, but we need a TODO fix here as the rand scalars should be at least 128bits (thus we should set vals[1] as well). Ideally I'd create a private function get_128bit_scalar and invoke that.

File: fastcrypto/src/tests/signature_tests.rs [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+use rand::{rngs::StdRng, SeedableRng};
+
+use crate::{
+    hash::{HashFunction, Sha256},
+    traits::KeyPair,
+};
+
+/// Generate `n` random key pairs.
+pub fn keys<KP: KeyPair>(n: usize) -> Vec<KP> {
```
Comment (kchalkias) [inline fastcrypto/src/tests/signature_tests.rs:12]: ah, these are util functions, not tests, so we could rename the file to signature_test_utils.rs

**Review by kchalkias — APPROVED** [review]
Some minor comments but logic wise we are golden! Thanks

**Comment by jonas-lj** [comment]
> Some minor comments but logic wise we are golden! Thanks

Great! Thanks for the review. I'll merge and add issues based on your comments.

**Pull request merged**